File: WordSearch.py
class Solution:
    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
    def exist(self, board, word):
        if not board or not word:
            return False
        for i in xrange(len(board)):
            board[i] = list(board[i])
        h, w = len(board), len(board[0])
        res = False
        def testValid(xx, yy):
            if xx < 0 or xx > h-1 or yy < 0 or yy > w-1:
                return False
            return True
            
        def dfs(x, y, w):
            if len(w) == 0:
                return True
            DIR = [(-1, 0), (1, 0), (0, -1), (0, 1)]
            for i, j in DIR:
                if testValid(x+i, y+j) and board[x+i][y+j] == w[0]:
                    tmp, board[x][y] = board[x][y], '#'
                    if dfs(x+i, y+j, w[1:]):
                        return True
                    board[x][y] = tmp
            
        for i in xrange(h):
            for j in xrange(w):
                if board[i][j] == word[0]:
                    if dfs(i,j,word[1:]):
                        return True
        return False

if __name__ == '__main__':
    so = Solution()
    bd = [
        "ABCE",
        "SFCS",
        "ADEE"
    ]
 
    w1 = "ABCCED"
    w2 = "SEE"
    w3 = "ABCB"

    bd2 = [
        "CAA",
        "AAA",
        "BCD"
    ]
    w4 = "AAB"
    print so.exist(bd2, w4)
        

File: CPP/Second/word_search.cc
// https://oj.leetcode.com/problems/word-search/

// Given a 2D board and a word, find if the word exists in the grid.
// The word can be constructed from letters of sequentially adjacent cell, where
// "adjacent" cells are those horizontally or vertically neighboring. The same
// For example,
// Given board =
// [
//   ["ABCE"],
//   ["SFCS"],
//   ["ADEE"]
// ]
// word = "ABCCED", -> returns true,
// word = "SEE", -> returns true,
// word = "ABCB", -> returns false.

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
private:
    int m, n;
public:
    bool exist(vector<vector<char> > &board, string word) {
        m = board.size(), n = board[0].size();
        vector<vector<char> > visited(m, vector<char>(n, false));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(board, word, i, j, visited))
                    return true;
            }
        }
        return false;
    }
    bool dfs(const vector<vector<char> > &board, string word, int x, int y, vector<vector<char> > &visited) {
        if (word.size() == 0) 
            return true;
        if (x < 0 || x >= m || y < 0 || y >= n)
            return false; 
        if (word[0] != board[x][y])
            return false;
        if (visited[x][y])
            return false;
        visited[x][y] = true;
        bool ret =  dfs(board, word.substr(1), x-1, y, visited) ||
                    dfs(board, word.substr(1), x+1, y, visited) ||
                    dfs(board, word.substr(1), x, y-1, visited) ||
                    dfs(board, word.substr(1), x, y+1, visited);
        visited[x][y] = false;
        return ret;
    }
};

int main(int argc, char *argv[])
{

    return 0;
}
